### PR TITLE
Create resource_test Suite

### PR DESCRIPTION
**Pull request by gabesaba**
#### What type of PR is this?
/kind cleanup
/kind feature

#### What this PR does / why we need it:
Add test suite to give more confidence in the new resource accounting logic (#2881). Part of #79

```release-note
NONE
```

### CODE REVIEW

File: pkg/cache/resource_test.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+					ResourceGroup(
+						*utiltesting.MakeFlavorQuotas("red").Resource("cpu", "10").Obj(),
+					).Cohort,
+			},
+			usage: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 31_000},
+			},
+			wantAvailable: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: -11000},
+				"cq2": {{Flavor: "red", Resource: "cpu"}: -1000},
```
Comment (alculquicondor) [inline pkg/cache/resource_test.go:141]: interesting....
What would happen if those 10 CPUs that CQ2 has are split into two CQs?
What would the usage on each of them be, and why?

File: pkg/cache/resource_test.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+			},
+			usage: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 20_000},
+				"cq2": {{Flavor: "red", Resource: "cpu"}: 10_000},
+				"cq3": {{Flavor: "red", Resource: "cpu"}: 5_000},
+			},
+			wantAvailable: map[string]resources.FlavorResourceQuantities{
+				"cq1": {{Flavor: "red", Resource: "cpu"}: 3_000},
+				"cq2": {{Flavor: "red", Resource: "cpu"}: 2_000},
+				"cq3": {{Flavor: "red", Resource: "cpu"}: 5_000},
```
Comment (alculquicondor) [inline pkg/cache/resource_test.go:184]: I think each of these values require a comment

File: pkg/cache/resource_test.go (hunk 1 above)
Comment (gabesaba) [reply pkg/cache/resource_test.go:141]: `cohort` SubtreeQuota is 20k (from `cohort`, and CQ1). Usage in cohort is 31k.  When checking usage, CQ2 and CQ3 would find 5k available locally and -11k in the parent. Adding those numbers would result in `available` -6k for each of those CQs.

Comment (gabesaba) [reply pkg/cache/resource_test.go:141]: When a negative number is returned, it indicates that some CQ/Cohort in the CohortTree is using too much resources - which can happen after a change of available resources, or switching parent Cohort.

This ensures that, for any CohortTree, we won't admit a workload that would result in using more of a FlavorResource than is in the CohortTree. I think it is desirable that this also affects CQs which do not lend capacity to or borrow capacity from the CohortTree - once we are overadmited, we should prevent any admissions which worsen the debt.

File: pkg/cache/resource_test.go (hunk 2 above)
Comment (gabesaba) [reply pkg/cache/resource_test.go]: Done. Also modified the numbers slightly to make the math more simple (3->5)

File: pkg/cache/resource_test.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/cache/resource_test.go:152]: As discussed offline, consider just returning zero when the CQ is oversubscribed.
IIUC, it shouldn't have any effect on algorithms, and this number is not reported to the user.
The advantage is just more maintainable unit tests.

**Review by alculquicondor — COMMENTED** [review]
/approve
/lgtm
/hold
(in case my suggestion is easily achievable)

File: pkg/cache/resource_test.go (hunk 1 above)
Comment (gabesaba) [reply pkg/cache/resource_test.go:152]: Done. We report the capacity lacked to the user during flavor assignment. Updated the overadmission integration test with a new number.

File: pkg/cache/resource_test.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+			for _, cq := range tc.clusterQueues {
+				_ = cache.AddClusterQueue(ctx, &cq)
+			}
+			for _, cohort := range tc.cohorts {
+				cache.AddCohort(&cohort)
+			}
+
+			snapshot := cache.Snapshot()
+
+			t.Run("before adding usage", func(t *testing.T) {
```
Comment (alculquicondor) [inline pkg/cache/resource_test.go:223]: remove `t.Run` here, as the remaining subtests depend on it.

Comment (gabesaba) [reply pkg/cache/resource_test.go]: Removed all 3, but left the blocks with a comment, so that the `got` variables are scoped.

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**